src/crimson/mgr/client.cc

```diff
@@ -78,16 +78,15 @@ void Client::ms_handle_connect(
   LOG_PREFIX(Client::ms_handle_connect);
   DEBUGDPP("", *this);
   ceph_assert_always(prv_shard == seastar::this_shard_id());
-  gates.dispatch_in_background(__func__, *this, [this, c] {
+  gates.dispatch_in_background(__func__, *this,
+  [this, c]() -> seastar::future<> {
     if (conn == c) {
       // ask for the mgrconfigure message
       auto m = crimson::make_message<MMgrOpen>();
       m->daemon_name = local_conf()->name.get_id();
       local_conf().get_config_bl(0, &m->config_bl, &last_config_bl_version);
       local_conf().get_defaults_bl(&m->config_defaults_bl);
-      return conn->send(std::move(m));
-    } else {
-      return seastar::now();
+      co_await conn->send(std::move(m));
```

Comment:
IIUC `seastar::coroutine::without_preemption_check` is needed in this `co_await` to make sure:
1. if (conn == c)
2. conn->send(std::move(m))

are atomic together (must be in the same seastar task)

Similar to other places are relying on checking the connection nullness atomically.

Reply:
Updated, I've omitted coroutine usage in the capturing lambdas.